controllers/disruption_controller.go

```diff
@@ -272,6 +329,97 @@ func (r *DisruptionReconciler) Reconcile(ctx context.Context, req ctrl.Request)
 	return ctrl.Result{}, nil
 }
 
+// createInstanceSelectorCache creates this instance's cache if it doesn't exist and attaches it to the controller
+func (r *DisruptionReconciler) manageInstanceSelectorCache(instance *chaosv1beta1.Disruption) error {
+	if !instance.Spec.DynamicTargeting {
+		return nil
+	}
+
+	disNamespacedName := types.NamespacedName{Name: instance.Name, Namespace: instance.Namespace}
+	disCompleteSelector, err := targetselector.GetLabelSelectorFromInstance(instance)
+
+	if err != nil {
+		return fmt.Errorf("error getting instance selector: %w", err)
+	}
+	// if it doesn't exist, create the cache context to re-trigger the disruption
+	if r.CachesCancel[disNamespacedName] == nil {
+		// create the cache/watcher with its options
+		var cacheOptions k8scache.Options
+		if instance.Spec.Level == chaostypes.DisruptionLevelNode {
+			cacheOptions = k8scache.Options{
+				SelectorsByObject: k8scache.SelectorsByObject{
+					&corev1.Node{}: {Label: disCompleteSelector},
+				},
+			}
+		} else {
+			cacheOptions = k8scache.Options{
+				SelectorsByObject: k8scache.SelectorsByObject{
+					&corev1.Pod{}: {Label: disCompleteSelector},
+				},
+				Namespace: instance.Namespace,
+			}
+		}
+
+		cache, err := k8scache.New(
+			ctrl.GetConfigOrDie(),
+			cacheOptions,
+		)
+
+		if err != nil {
+			return fmt.Errorf("cache gen error: %w", err)
+		}
+
+		// attach handler to cache in order to monitor the cache activity
+		var info k8scache.Informer
+
+		if instance.Spec.Level == chaostypes.DisruptionLevelNode {
+			info, err = cache.GetInformer(context.Background(), &corev1.Node{})
+		} else {
+			info, err = cache.GetInformer(context.Background(), &corev1.Pod{})
+		}
+
+		if err != nil {
+			return fmt.Errorf("cache gen error: %w", err)
+		}
+
+		info.AddEventHandler(DisruptionSelectorHandler{disruption: instance, reconciler: r})
+
+		// start the cache with a cancelable context and attach it to the controller as a watch source
+		ch := make(chan error)
+		cacheCtx, cacheCancelFunc := context.WithCancel(context.TODO())
```

Comment:
I think it may be helpful to set a timeout here, equivalent to a bit longer than the disruption duration + GC. Right now the only way we remove these caches is upon working disruption deletion, but that isn't technically _guaranteed_ to happen (though it does in the 99.9% case, I know). After the timeout, we can just go ahead and cancel the cache and delete it.

Reply:
This seems fair. However I need to find a way to keep the sync between the context delete and the cancel func array cleanup. I'll find something.

Reply:
I just pushed a fix; on top of the timeout, the reconciler will now store both the cacheContext and its cancel function; if [`cacheContext.Err()`](https://pkg.go.dev/context#Context) is not nil, the map entry will be deleted. It seems bit heavy on the reconcile loop but I see no other solution than to check either periodically or with every reconcile.
Checking periodically (i.e. like the metrics) would add a new timing logic we haven't introduced before and I'm afraid it would be very unclear.
Obviously open to debate this if there's a disagreement.